iceoryx2-ffi/cxx/include/iox2/waitset.hpp

```diff
@@ -0,0 +1,222 @@
+///
+/// Can be created via the [`WaitSetBuilder`].
+template <ServiceType S>
+class WaitSet {
+  public:
+    WaitSet(const WaitSet&) = delete;
+    auto operator=(const WaitSet&) -> WaitSet& = delete;
+    WaitSet(WaitSet&&) noexcept;
+    auto operator=(WaitSet&&) noexcept -> WaitSet&;
+    ~WaitSet();
+
+    /// Can be called from within a callback during [`WaitSet::run()`] to signal the [`WaitSet`]
+    /// to stop running after this iteration.
+    void stop();
+
+    /// Waits in an infinite loop on the [`WaitSet`]. The provided callback is called for every
+    /// attachment that was triggered and the [`AttachmentId`] is provided as an input argument to
+    /// acquire the source.
+    /// If an interrupt- (`SIGINT`) or a termination-signal (`SIGTERM`) was received, it will exit
+    /// the loop and inform the user via [`WaitSetRunResult`].
+    auto run(const iox::function<void(AttachmentId<S>)>& fn_call) -> iox::expected<WaitSetRunResult, WaitSetRunError>;
+
+    /// Tries to wait on the [`WaitSet`]. The provided callback is called for every attachment that
+    /// was triggered and the [`AttachmentId`] is provided as an input argument to acquire the
+    /// source.
+    /// If nothing was triggered the [`WaitSet`] returns immediately.
+    auto run_once(const iox::function<void(AttachmentId<S>)>& fn_call) -> iox::expected<void, WaitSetRunError>;
+
+    /// Returns the capacity of the [`WaitSet`]
+    auto capacity() const -> uint64_t;
+
+    /// Returns the number of attachments.
+    auto len() const -> uint64_t;
+
+    /// Returns true if the [`WaitSet`] has no attachments, otherwise false.
+    auto is_empty() const -> bool;
+
+    /// Attaches a [`Listener`] as notification to the [`WaitSet`]. Whenever an event is received on the
+    /// object the [`WaitSet`] informs the user in [`WaitSet::run()`] to handle the event.
+    /// The object cannot be attached twice and the
+    /// [`WaitSet::capacity()`] is limited by the underlying implementation.
+    ///
+    /// # Safety
+    ///
+    /// * The [`Listener`] must life at least as long as the returned [`Guard`].
+    /// * The [`Guard`] must life at least as long as the [`WaitsSet`].
+    auto attach_notification(const Listener<S>& listener) -> iox::expected<Guard<S>, WaitSetAttachmentError>;
+
+    /// Attaches a [`FileDescriptorView`] as notification to the [`WaitSet`]. Whenever an event is received on the
+    /// object the [`WaitSet`] informs the user in [`WaitSet::run()`] to handle the event.
+    /// The object cannot be attached twice and the
+    /// [`WaitSet::capacity()`] is limited by the underlying implementation.
+    ///
+    /// # Safety
+    ///
+    /// * The corresponding [`FileDescriptor`] must life at least as long as the returned [`Guard`].
+    /// * The [`Guard`] must life at least as long as the [`WaitsSet`].
+    auto attach_notification(FileDescriptorView file_descriptor) -> iox::expected<Guard<S>, WaitSetAttachmentError>;
```

Comment:
If the goal was to just have a consistent API for attaching "things", I would just call it `attach` and let the argument inform the user of what is actually being attached.

Reply:
I would have named the Rust API the same, for the same reasons. If one of the goals is user-friendliness, we should change it. It can be done separately if you want to focus on the functionality here, but if we do that I worry it will just not get done.

Reply:
In Rust it would not work due to not supporting function overloading.

Reply:
@elBoberido Ah right, my brain still lives partially in C++ world.